File: deck/base/code/native/swift/runtime/websocket.swift
// WebSocket client runtime over URLSessionWebSocketTask. The opaque handle a seed socket holds is the task itself, which
// buffers incoming messages, so receive pulls the next one with no message-loss race. receive returns the seed `message`
// form. Reached only through the public WebSocket API.
import Foundation

enum websocket {
    static func connect(_ url: String) async -> URLSessionWebSocketTask {
        let task = URLSession.shared.webSocketTask(with: URL(string: url)!)
        task.resume()
        return task
    }

    static func send(_ task: URLSessionWebSocketTask, _ data: String) async {
        try? await task.send(.string(data))
    }

    static func receive(_ task: URLSessionWebSocketTask) async -> Message {
        do {
            let message = try await task.receive()
            switch message {
            case .string(let text):
                return Message(kind: "text", data: text)
            case .data:
                return Message(kind: "text", data: "")
            @unknown default:
                return Message(kind: "close", data: "")
            }
        } catch {
            return Message(kind: "close", data: "")
        }
    }

    static func close(_ task: URLSessionWebSocketTask) async {
        task.cancel(with: .normalClosure, reason: nil)
    }
}

File: deck/seed/code/native/node/network/runtime/websocket.ts
// WebSocket client runtime over node's global WebSocket (the browser-style API, stable since node 22). The opaque handle
// a seed socket holds is a small wrapper that buffers incoming frames: a permanent listener attached at connect time
// queues every message (and the close), and receive pulls the next one or waits. This makes the request / response
// pattern (send then receive) race-free, since a frame arriving before receive is called is held, not dropped. Reached
// only through the public WebSocket API.
type WebSocketFrame = { kind: string; data: string }
type WebSocketHandle = {
  socket: WebSocket
  queue: WebSocketFrame[]
  waiters: ((frame: WebSocketFrame) => void)[]
}

const websocket = {
  connect: (url: string): Promise<WebSocketHandle> =>
    new Promise((ok, fail) => {
      const socket = new WebSocket(url)
      const handle: WebSocketHandle = { socket, queue: [], waiters: [] }
      const deliver = (frame: WebSocketFrame) => {
        const waiter = handle.waiters.shift()
        if (waiter) waiter(frame)
        else handle.queue.push(frame)
      }
      socket.addEventListener('message', event =>
        deliver({ kind: 'text', data: String(event.data) }),
      )
      socket.addEventListener('close', () =>
        deliver({ kind: 'close', data: '' }),
      )
      socket.addEventListener('open', () => ok(handle))
      socket.addEventListener('error', () =>
        fail(new Error('websocket connect failed')),
      )
    }),

  send: (handle: WebSocketHandle, data: string): Promise<void> => {
    handle.socket.send(data)
    return Promise.resolve()
  },

  receive: (handle: WebSocketHandle): Promise<WebSocketFrame> =>
    new Promise(ok => {
      const frame = handle.queue.shift()
      if (frame) ok(frame)
      else handle.waiters.push(ok)
    }),

  close: (handle: WebSocketHandle): Promise<void> => {
    handle.socket.close()
    return Promise.resolve()
  },
}
